### ISSUE TITLE
LoopDefs do not support segments between loops

### ISSUE THREAD

**Issue by jakeonfire** [kputnam/stupidedi#97]
I'm trying to implement the BE 834 spec (Benefit Enrollment and Maintenance) and am having a couple issues. For one, within one of the loops there are header segments, child loops, then another segment to indicate whether a final loop (or multiple) is present, and LoopDef cannot currently handle this format. I noticed there are some other incomplete specs (e.g. HB271) which also have this structure (see loop 2120). Any suggestions? I'm happy to help contribute...

Also, I'm having the same issue as https://github.com/kputnam/stupidedi/issues/87 where, despite a TransactionSetDef, an error is raised about a specific implementation of it missing - is a Guide also required in that case? It might be nice to update `Transcribing.md` with that info if so...

**Comment by kputnam** [comment]
Hi Jake, let me get back to you on this over the weekend. I recall LoopDef isn't designed to have segments between child loops, and more importantly, the parser wouldn't know what to do with them, but I'll have to spend some time to reacquaint myself before I can offer any suggestions.

Regarding the second problem, you can use a TransactionSetDef as long as it's wired into the `Config` -- maybe you used the default construct and it doesn't include your custom spec? You can see how the HIPAA specs were added here (both `TransactionSetDef`s and their corresponding `Guide`s):
    https://github.com/kputnam/stupidedi/blob/master/lib/stupidedi/config.rb#L97

**Comment by kputnam** [comment]
I haven't looked at BE 834 yet, but it might be possible to make a grammar that deviates a bit from the actual specification. This will be much easier than trying to change the parser to handle this (which would be very difficult and risky to do at this point, I think).

For instance, in the case of HB 271, you could move the LS segment to be an optional segment just before the 2115 loop (to handle the case when 2115 does not occur in the input) and also copied as an optional trailer segment at the bottom of the 2115 loop (to handle the case when the 2115 does occur). You might have to change the sequence number of the first LS to be above 2500 and below 2600 for this to work.

```ruby
                    s::SD1.use(2300, r::Optional,  d::RepeatCount.bounded(1)),
                    s::PKD.use(2400, r::Optional,  d::RepeatCount.bounded(1)),
                    s::MSG.use(2500, r::Optional,  d::RepeatCount.bounded(10)),
                    s:: LS.use(2550, r::Optional,  d::RepeatCount.bounded(1)), ## here

                    d::LoopDef.build("2115", d::RepeatCount.unbounded,
                      s::III.use(2600, r::Optional,  d::RepeatCount.bounded(1)),
                      s::DTP.use(2700, r::Optional,  d::RepeatCount.bounded(5)),
                      s::AMT.use(2800, r::Optional,  d::RepeatCount.bounded(5)),
                      s::PCT.use(2900, r::Optional,  d::RepeatCount.bounded(5)),

                      d::LoopDef.build("2117", d::RepeatCount.unbounded,
                        s:: LQ.use(3000, r::Optional,  d::RepeatCount.bounded(1)),
                        s::AMT.use(3100, r::Optional,  d::RepeatCount.bounded(5)),
                        s::PCT.use(3200, r::Optional,  d::RepeatCount.bounded(5))),

                      s:: LS.use(3300, r::Optional,  d::RepeatCount.bounded(1))), ## here

                    d::LoopDef.build("2120", d::RepeatCount.unbounded,
                      s::NM1.use(3400, r::Optional,  d::RepeatCount.bounded(1)),
                      s:: N2.use(3500, r::Optional,  d::RepeatCount.bounded(1)),
```

This would accept any input that the original grammar would have, plus some additional "wrong" inputs that had extra LS segments. Personally, I wouldn't be too concerned that a trading partner could send exactly the wrong input to trick the parser into thinking it's correct, nor would I be too concerned that I would construct that myself (just review your code carefully); but you could write a program to check for extra LS segments if you were worried about that.

I'll check to see if I have the documentation for BE 834 tomorrow. Hopefully this kind of trick will also work for that.

**Comment by jakeonfire** [comment]
I'll try this out on Wednesday. Sounds like it should work.

**Comment by jakeonfire** [comment]
in the case of BE 834 there are 9 optional loops before the final optional segment [LS] and loop [2710], so this pattern requires 10 misplaced LS segments (the first at the end of the last required loop [2100A] and the others at the end of the following 9 optional loops [2100B - 2100H, 2200, and 2300]). That's quite a few...

Also, the detail header's first loop is unbounded in the spec, but that raises an error:
```
2000 - MEMBER LEVEL DETAIL is an entry point into Table 2 - Detail. Both are marked repeatable, but only one should be repeatable to avoid generating an ambiguous grammar. You probably want to change the repeat count of 2000 - MEMBER LEVEL DETAIL (Stupidedi::Exceptions::InvalidSchemaError)
```
If i change it to `bounded(1)` it works but that also feels like a hack. Why is it assumed that `TableDef.detail` is repeatable?

**Comment by jakeonfire** [comment]
I fixed the segment-between-loops issue by adding another loop (which is kind of an implied loop anyway, it's part of the 'Standard'), and that feels better than 10 misplaced segments. I also added an optional, last, named parameter to TableDef.detail to allow setting `repeat: false` (the default is `true`). And now it seems to work! I ran `edi-ed` against the samples and they seem to work. I'll take a look at the specs and then make a pull-request for adding BE 834 😄
